### PR TITLE
[RESTEASY-2249] Fixing private PostConstruct method detection

### PR DESCRIPTION
**Pull request by spyrkob**
JIRA: https://issues.jboss.org/browse/RESTEASY-2249

@ronsigal I think there's also a bug in code detecting Interceptor PostConstruct methods in JaxrsInjectionTarget - it only checks for parameters, but the spec allows those methods to have a return type and throw a checked exception as well. I tried fixing that, but I was not able to get your EJB test in PostConstructInjectionTest to work (currently TestPostInjectCdiOn only checks /normal/get path never /ejb/get)

### CODE REVIEW

File: resteasy-core-spi/src/main/java/org/jboss/resteasy/spi/util/Types.java [hunk 1]
```diff
@@ -810,7 +810,7 @@ public static boolean hasPostConstruct(Class<?> clazz)
 
    public static boolean hasPostConstruct(Class<?> clazz, Function<Method, Boolean> validateParameterCount)
    {
-      for (Method m : clazz.getMethods())
+      for (Method m : clazz.getDeclaredMethods())
```
Comment (ronsigal) [inline resteasy-core-spi/src/main/java/org/jboss/resteasy/spi/util/Types.java]: Hmmm. You're right, getMethods() returns only public methods. On the other hand, getDeclaredMethods() doesn't look at inherited methods. Probably needs to call getDeclaredMethods() while recursively visiting all superclasses. Thank you.

**Comment by ronsigal** [comment]
Hi @spyrkob,

I'm thinking I might have looked at https://docs.oracle.com/javaee/5/api/javax/annotation/PostConstruct.html, which says "The return type of the method MUST be void", instead of the newer https://docs.oracle.com/javaee/7/api/javax/annotation/PostConstruct.html.

-Ron

**Comment by ronsigal** [comment]
re: "I was not able to get your EJB test in PostConstructInjectionTest to work (currently TestPostInjectCdiOn only checks /normal/get path never /ejb/get)"

Ah, yes. Thanks again. I'll have to look into that.

File: resteasy-core-spi/src/main/java/org/jboss/resteasy/spi/util/Types.java (hunk 1 above)
Comment (spyrkob) [reply resteasy-core-spi/src/main/java/org/jboss/resteasy/spi/util/Types.java]: Thanks @ronsigal I didn't think of that.
Added a recursive check for superclasses, if it looks good, I'll squash the two commits

Comment (ronsigal) [reply resteasy-core-spi/src/main/java/org/jboss/resteasy/spi/util/Types.java]: So far, so good, @spyrkob. Thanks.

There's also the alternative signatures for Interceptors, but I still have to look at the EJB case.

**Comment by ronsigal** [comment]
Well, @spyrkob, I've learned two things about the EJB test.
1. It makes no sense the way it is written. The @PostConstruct method should be in an interceptor.
2. When the test is fixed, it fails because validation called from JaxrsInjectionTarget.inject() occurs before the call to the @PostConstruct method.

I've created RESTEASY-2264 "Validation and @PostConstruct, Part II" with an updated version of PostConstructInjectionTest. @spyrkob, could you update PostConstructInjectionTest and related files in your pull request? Thanks.

**Comment by spyrkob** [comment]
Added the test changes from RESTEASY-2264 with disabled EJB test. Lemme know if it looks all right, and I'll squash the commits.

**Comment by spyrkob** [comment]
@ronsigal I forgot to follow up on this, sorry. Does the PR look good to be merged in?

**Comment by ronsigal** [comment]
Hi @spyrkob,

I also forgot all about it. Thank you for cleaning up the EJB mess I made. It looks good. In the interest of best practices, yes, could you squash the commits?

Also, do you feel like doing a PR for branch 3.10.0.Final?

-Ron

**Comment by spyrkob** [comment]
Thanks @ronsigal 
Ports for 3.10 (#2164) and 3.6.1 (#2161) are up.

**Pull request merged**